core/src/main/java/jenkins/util/SystemProperties.java

```diff
@@ -60,40 +74,112 @@
-         */
-        @CheckForNull
-        private static ServletContext theContext;
-
         /**
          * Called by the servlet container to initialize the {@link ServletContext}.
          */
         @Override
-        @SuppressFBWarnings(value = "ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD",
-                justification = "Currently Jenkins instance may have one ond only one context")
         public void contextInitialized(ServletContextEvent event) {
-            theContext = event.getServletContext();
+            ServletContext theContext = event.getServletContext();
+            handler = new Handler() {
+                @Override
+                public String getString(String key) {
+                    if (StringUtils.isNotBlank(key)) {
+                        try {
+                            return theContext.getInitParameter(key);
+                        } catch (SecurityException ex) {
+                            // Log exception and go on
+                            LOGGER.log(Level.CONFIG, "Access to the property {0} is prohibited", key);
+                        }
+                    }
+                    return null;
+                }
+                @Override
+                public Map<String, String> getAllStrings() {
+                    Map<String, String> values = new HashMap<>();
+                    Enumeration<String> names = theContext.getInitParameterNames();
+                    while (names.hasMoreElements()) {
+                        String key = names.nextElement();
+                        String value = getString(key);
+                        if (value != null) {
+                            values.put(key, value);
+                        }
+                    }
+                    return values;
+                }
+            };
         }
 
         @Override
         public void contextDestroyed(ServletContextEvent event) {
-            theContext = null;
+            handler = NULL_HANDLER;
         }
 
     }
 
+    @Extension
+    public static final class SlaveCopier extends ComputerListener {
+        @Override
+        public void preOnline(Computer c, Channel channel, FilePath root, TaskListener listener) throws IOException, InterruptedException {
+            channel.call(new CopySystemProperties());
+        }
+        private static final class CopySystemProperties extends MasterToSlaveCallable<Void, RuntimeException> {
+            private static final long serialVersionUID = 1;
+            private final Map<String, String> allStrings;
+            @SuppressWarnings("unchecked")
+            CopySystemProperties() {
+                // Take a snapshot of all system properties and context variables available on the master at the time the agent starts.
+                allStrings = new HashMap<>();
+                allStrings.putAll(handler.getAllStrings());
+                allStrings.putAll((Map) System.getProperties()); // these take precedence
```

Comment:
If I understand correctly, this would mean that calling `SystemProperties.getString("file.separator")` on the agent would actually get you the value from the master, which is confusing at best but could also be seriously problematic if the agent is running a different OS than the master. Not necessarily a huge problem as long as anyone using `SystemProperties` is careful to avoid those variables, and the API is `@Restricted(NoExternalUse.class)`, but maybe it would be safer to only copy properties whose keys contain `hudson` or `jenkins` in the name?

Reply:
> calling `SystemProperties.getString("file.separator")` on the agent would actually get you the value from the master

No, because `getString` will first check the _local_ system properties. This snapshot is only used for fallbacks of variables not set on the agent.

Reply:
I don't think this should be a problem, since even on master you would get weird results today if part of your code relied on the System property value alone for a given key, while the rest takes into account the context variable. If the former is undefined, `SystemProperties` might still have a value but `System.getProperty` won't.